docs/design/git-submodule-storage.md

```diff
@@ -59,22 +59,88 @@ These extensions are:
 - Colocated Git repos
 - Non-git backends
 
-## Possible approaches
+## Proposed design
 
-### Approach 1: Store Git submodules as full jj repos
+Git submodules will be stored as full jj repos. In the code, jj commands will
+only interact with the submodule's repo as an entire unit, e.g. it cannot query
+the submodule's commit backend directly. A well-abstracted submodule will extend
+well to non-git backends and non-git subrepos.
 
-This would be somewhere in `.jj` but outside of `.jj/store`. We would then
-expose a "submodules" interface that gets hooked up to the relevant machinery
-(e.g. updating the working copy).
+The main challenge with this approach is that the submodule repo can be in a
+state that is internally valid (when consider only the submodule's repo), but
+invalid when considering the superproject-submodule system. This will be managed
+by requiring all submodule interactions go through the superproject so that
+superproject-submodle coordination can occur. For example, jj will not allow the
+user to work on the submodule's repo without going through the superproject
+(unlike Git).
 
-TODO(chooglen): Discuss operation log
-TODO(chooglen): Discuss nested submodules
+The notable workflows could be addressed like so:
 
-### Approach 3: Store Git submodules as alternate jj repo backends
+### Fetching submodule commits
+
+The submodule would fetch using the equivalent of `jj git fetch`. It remains to
+be decided how a "recursive" fetch should work, especially if a newly fetched
+superproject commit references an unfetched submodule commit. A reasonable
+approximation would be to fetch all branches in the submodule, and then, if the
+submodule commit is still missing, gracefully handle it.
+
+### "jj op restore" and operation log format
+
+As full repos, each submodule will have its own operation log. We will continue
+to use the existing operation log format, where each operation log tracks their
+own repo's commits. As commands are run in the superproject, corresponding
+commands will be run in the submodule as necessary, e.g. checking out a
+superproject commit will cause a submodule commit to also be checked out.
+
+Since there is no association between a superproject operation and a submodule
+operation, `jj op restore` in the superproject will not restore the submodule to
+a previous operation. Instead, the appropriate submodule operation(s) will be
+created. This is sufficient to preserve the superproject-submodule relationship;
+it precludes "recursive" restore (e.g. restoring branches in the superproject
+and submodules) but it seems unlikely that we will need such a thing.
```

Comment:
That seems fine to start with. It's probably not very hard to add pointers to submodule operations if we decide that that's useful.